src/share/classes/com/sun/javatest/regtest/agent/MainActionHelper.java

```diff
@@ -247,10 +258,11 @@ public AStatus runClass() {
                 }
             }
 
-            if (((avmr.t != null) || (tg.uncaughtThrowable != null)) && (error == null)) {
+            if (((avmr.t != null) || (tg.uncaughtThrowable != null) || (globalUncaughtThrowable != null))
+                    && (error == null)) {
                 if (avmr.t == null) {
-                    error = tg.uncaughtThrowable;
-                } else {
+                    error = tg.uncaughtThrowable != null ? tg.uncaughtThrowable : globalUncaughtThrowable;
```

Comment:
Are you missing synchronisation here? The VM-side UHE is set while holding the lock for the MainActionHelper instance whereas help you read without any synchronisation. If you make globalUncaughtThrowable volatile then it would be okay, meaning keep the synchronisation in the setter code, read from elsewhere without synchronisation.

Reply:
Thanks, added volatile.